flocker/node/agents/cinder.py

```diff
@@ -0,0 +1,114 @@
+* I haven't tried the standard docs.openstack.org/developer/python-cinderclient library, but I guess that *should* be the best option.
+  It might be worth experimenting with that a little, especially to see how it handle volume metadata.
+"""
+
+import pyrax
+
+from zope.interface import implementer
+
+from .blockdevice import IBlockDeviceAPI
+
+
+@implementer(IBlockDeviceAPI)
+class CinderBlockDeviceAPI(object):
+    """
+    A cinder implementation of ``IBlockDeviceAPI`` which creates block devices
+    in an OpenStack cluster.
+    """
+    def __init__(self, cluster_id, region):
+        """
+        :param UUID cluster_id: An ID that will be included in the names of
+            Cinder block devices in order to associate them with a particular
+            Flocker cluster.
+        :param unicode region: A provider specific region identifier string.
+        :param pyrax_context: An authenticated pyrax context.
+        """
+
+    def create_volume(self, dataset_id, size):
+        """
+        Create the block device using the volume_driver.
+        Store the dataset_id as metadata
+        Store the cluster_id as metadata
+        Assign a Human readable name and description?
+        Eg FLOCKER: Block device for dataset {dataset_id} in cluster {cluster_id}
+
+        ```
+        In [10]: volume.create('richardw-test-2', 100, 'SATA', metadata={"flocker_dataset_id": unicode(uuid.uuid4())})
+        Out[10]: <CloudBlockStorageVolume attachments=[], availability_zone=nova, bootable=false, created_at=2015-04-17T11:59:48.475838, display_description=, display_name=richardw-test-2, id=c7b08f98-f363-484d-83fb-410927c69159, metadata={u'flocker_dataset_id': u'b816da15-063e-47a4-843d-275ffa37ecec'}, size=100, snapshot_id=None, source_volid=None, status=creating, volume_type=SATA>
+        ```
+
+        ONLY TROUBLE IS...that the metadata I supplied doesn't seem to be returned when I then call ``list()`` :-(
+
+        Not sure if that's a problem with pyrax or whether it's simply not possible to store custom metadata for volumes on Rackspace.
+
+        TODO: Tom suggested asking Rackspace directly.
+
+        If it's not possible, we can maybe store the dataset_id etc as JSON in the volume description field...but that sounds awful.
+
+        Return a BlockDeviceVolume via _blockdevicevolume_from_pyrax_volume
+
+        http://docs.rackspace.com/cbs/api/v1.0/cbs-devguide/content/POST_createVolume_v1__tenant_id__volumes_volumes.html
+
+        Discussion:
+         * Rackspace (maybe cinder in general) supports a block device type, eg SSD or SATA.
+           I guess we'll hardcode SATA here to start with
+           Maybe expand the API later to make the disk type configurable?
+
+         * pyrax.volume.create expects a size in GB
+           The minimum SATA disk size on Rackspace is 100GB.
+           How do we enforce that?
+           And what (if any) errors should we raise if the user requests something smaller?
```

Comment:
Some useful, informative exception should be raised.  We won't be able to do anything but log it for now.  Later, the HTTP API should prevent bad values from getting this far.  And later still, there will be a way to discover convergence errors via the HTTP API.